py-datatrove: add new package and dependencies

thomas-bouvier opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

I started working on this at the old location [here](https://github.com/spack/spack/pull/50737). Resubmitting a PR with additional fixes.

thomas-bouvier committed:
py-datatrove: add package and dependencies

thomas-bouvier committed:
kenlm: fix build

thomas-bouvier committed:
[@spackbot] updating style on behalf of thomas-bouvier

alecbcs commented:
Giving the Spack Tutorial today. I'll take a look at this tomorrow :)

wdconinc commented on repos/spack_repo/builtin/packages/py_orjson/package.py:
```diff
@@ -23,11 +24,11 @@ class PyOrjson(PythonPackage):
-            depends_on("rust@1.60:")
-            depends_on("python@3.7:")
-            depends_on("py-maturin@0.13:0.14")
         with when("@03.9:"):
```
```suggestion
         with when("@3.9:"):
```

wdconinc commented on repos/spack_repo/builtin/packages/py_httptools/package.py:
```diff
@@ -15,10 +15,12 @@ class PyHttptools(PythonPackage):
     depends_on("c", type="build")  # generated
 
     depends_on("py-setuptools", type="build")
+    depends_on("py-cython@0.29.24:", when="@0.6.4:", type="build")
     depends_on("py-cython@0.29.24:0.29", type="build")
```
If the intention is to open this up beyond 0.29, then this construction doesn't do that since the second constraint remains active for 0.6.4 and newer.

wdconinc reviewed (commented):
I'll try to look in more detail later. Some quick comments.

wdconinc commented on repos/spack_repo/builtin/packages/kenlm/package.py:
```diff
@@ -0,0 +1,45 @@
+    version("master", branch="master")
+
+    variant("python", default=True, description="Build Python bindings")
+    variant("debug", default=False, description="Build with debug flags")
```
CMakePackages automatically have `build_type [Release]      Debug, MinSizeRel, RelWithDebInfo, Release`. This is duplication.

thomas-bouvier replied:
Thank you for the reminder. I am not sure what to do regarding the Python bindings though.

wdconinc replied:
The python variant to enable python bindings is fine and an intended use of variants.

wdconinc commented on repos/spack_repo/builtin/packages/py_fastapi/package.py:
```diff
@@ -36,6 +37,7 @@ class PyFastapi(PythonPackage):
     depends_on("py-hatchling", when="@:0.110.2", type="build")
 
+    depends_on("py-starlette@0.40.0:0.46", when="@0.115.12:", type=("build", "run"))
     depends_on("py-starlette@0.40:0.41", when="@0.115.3:", type=("build", "run"))
```
This will remain restricted to `:0.41`, even for `0.115.12`.

thomas-bouvier committed:
py-datatrove: address reviews

thomas-bouvier commented:
Thank you @wdconinc for the review. Everything compiles correctly on my end.

wdconinc reviewed (dismissed):
Looks good to me. But I didn't try to install and test, or verify all the dependency version changes.

bernhardkaindl reviewed (dismissed):
Summary / the last messages:

@thomas-bouvier 
> Thank you wdconinc for the review. Everything compiles correctly on my end.

wdconinc:
> Looks good to me. But I didn't try to install and test, or verify all the dependency version changes.

Confirming, it builds on Ubuntu:
```py
-- linux-ubuntu22.04-skylake / no compilers ---------------------
py-datatrove@0.5.0+io+processing build_system=python_pip
```

bernhardkaindl reviewed (commented):
@thomas-bouvier, you need to resolve the conflicts.

The changes to pytest have been merged with https://github.com/spack/spack-packages/pull/862 into spack already, the change to it has to be removed from this PR.

Otherwise, the conflict to repos/spack_repo/builtin/packages/py_orjson/package.py needs to be resolved.

thomas-bouvier commented:
Thank you @bernhardkaindl. I rebased and fixed merge conflicts.

@wu-kan, based on https://github.com/ijl/orjson/blob/3.10.3/pyproject.toml I don't understand the following constraint that you added a few days ago in `py_orjson/packages.py`:

```python
    with when("@:3.10.3"):
        depends_on("python@:3.12")
```

bernhardkaindl commented on repos/spack_repo/builtin/packages/py_httptools/package.py:
```diff
@@ -21,6 +21,9 @@ class PyHttptools(PythonPackage):
-    depends_on("py-cython@0.29.24:0.29", type="build", when="@:0.6.1")
+    depends_on("py-cython@0.29.24:", when="@0.6.2:", type="build")
+    depends_on("py-cython@0.29.24:0.29", when="@0.5", type="build")
+    depends_on("py-cython@0.29.24", when="@0.1", type="build")
```
The fixed dependency `"py-cython@0.29.24", when="@0.1"` maybe ok as 0.1 is likely a very old version.

bernhardkaindl commented on repos/spack_repo/builtin/packages/kenlm/package.py:
```diff
@@ -0,0 +1,39 @@
+        env.set("CXXFLAGS", "-fPIC")
+
+    def install(self, spec, prefix):
+        # Kenlm lacks an install in its CMakeList
+        install_tree(join_path(self.build_directory, "bin"), prefix.bin)
```
Copilot may have a point here, as when
```py
 variant("python", default=True, description="Build Python bindings")
```
the python bindings would usually not be in the `bin/` directory.
Datatrove likely does not need them, so I guess the python bindings could be disabled by default, and a TODO could be added if anyone wants them.

thomas-bouvier replied:
I don't know how to proceed. From my understanding, py-datatrove uses the Python bindings, and the current recipe allows me to `import kenlm`.

thomas-bouvier replied:
@bernhardkaindl how should I proceed with this?

wdconinc replied:
Why do you implement the `install` method at all? It seems that there is install info in the CMakeLists.txt, e.g. https://github.com/kpu/kenlm/blob/4cb443e60b7bf2c0ddf3c745378f76cb59e254e5/CMakeLists.txt#L131.

In any case, copilot is incorrect on the location of the python site-packages. Those should go in the package prefix, not the python dependency prefix.

wdconinc replied:
Why do you implement the `install` method at all? It seems that there is install info in the CMakeLists.txt, e.g. https://github.com/kpu/kenlm/blob/4cb443e60b7bf2c0ddf3c745378f76cb59e254e5/CMakeLists.txt#L131.

In any case, copilot is incorrect on the location of the python site-packages. Those should go in the package prefix, not the python dependency prefix.

thomas-bouvier replied:
At some point I had an error saying that Spack couldn't detect the install target, but you're right, I can't reproduce that behavior anymore. I removed the reimplementation.

wu-kan commented:
> Thank you @bernhardkaindl. I rebased and fixed merge conflicts.
> 
> @wu-kan, based on https://github.com/ijl/orjson/blob/3.10.3/pyproject.toml I don't understand the following constraint that you added a few days ago in `py_orjson/packages.py`:
> 
> ```python
>     with when("@:3.10.3"):
>         depends_on("python@:3.12")
> ```

It failed on my machine on `@:3.10.3 ^ python@3.13`

<img width="3683" height="1813" alt="screenshot-20250807-105614" src="https://github.com/user-attachments/assets/5fd0b6d3-fcc6-402a-a419-0a25eea41dc4" />

succeed on `@:3.10.3 ^ python@:3.12`

<img width="2858" height="645" alt="screenshot-20250807-105921" src="https://github.com/user-attachments/assets/cae2f450-3ecc-4104-8dac-7afbab535572" />

succeed on `@3.11.1 ^ python@3.13`

<img width="2585" height="670" alt="screenshot-20250807-110156" src="https://github.com/user-attachments/assets/af65482c-9712-4d91-ae43-3f6c0ca2363e" />

thomas-bouvier commented on repos/spack_repo/builtin/packages/py_httptools/package.py:
```diff
@@ -21,6 +21,9 @@ class PyHttptools(PythonPackage):
-    depends_on("py-cython@0.29.24:", type="build", when="@0.6.2:")
-    depends_on("py-cython@0.29.24:0.29", type="build", when="@:0.6.1")
+    depends_on("py-cython@0.29.24:", when="@0.6.2:", type="build")
+    depends_on("py-cython@0.29.24:0.29", when="@0.5", type="build")
```
I think you meant version @0.1? In this case, the dependency is on cython v0.29.14 precisely https://github.com/MagicStack/httptools/blob/v0.1.1/setup.py (which I fixed because it was actually set to 0.29.24--my mistake).

thomas-bouvier committed:
py-httptools: fix dependency version

thomas-bouvier commented:
I need some guidance on whether the kenlm recipe needs fixing—on my end, everything is working perfectly as is.

tldahlgren commented on repos/spack_repo/builtin/packages/kenlm/package.py:
```diff
@@ -0,0 +1,39 @@
+    extends("python", when="+python")
+
+    depends_on("c", type="build")  # generated
+    depends_on("cxx", type="build")  # generated
```
If language dependencies have been confirmed then the `# generated` comments should be removed.

thomas-bouvier replied:
Fixed in latest commit

tldahlgren replied:
Ah. that must've been my pending comment GitHub referenced when I was prompting me to finish my review earlier.

thomas-bouvier committed:
kenlm: fix Python build

thomas-bouvier commented:
@spackbot fix style

thomas-bouvier committed:
[@spackbot] updating style on behalf of thomas-bouvier

thomas-bouvier commented:
Many thanks for the review @wdconinc. I fixed the kenlm build.

`py-httptools@0.1.1` depends on `py-cython@0.29.14` which is not in Spack. Can I safely remove `py-httptools@0.1.1`? Only `py-uvicorn` depends on `py-httptools@0.5:`, so that should not be a problem as no package needs `py-httptools@0.1.1`.

tldahlgren reviewed (commented):
Confirmed homepages, branch and 17 sha256s.

tldahlgren commented on repos/spack_repo/builtin/packages/py_courlan/package.py:
```diff
@@ -0,0 +1,23 @@
+    version("1.3.2", sha256="0b66f4db3a9c39a6e22dd247c72cfaa57d68ea660e94bb2c84ec7db8712af190")
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
```
Per https://github.com/adbar/courlan/blob/v1.3.2/pyproject.toml:


```suggestion
    depends_on("py-setuptools@61:", type="build")
```

tldahlgren commented on repos/spack_repo/builtin/packages/py_datatrove/package.py:
```diff
@@ -0,0 +1,61 @@
+    depends_on("py-rich", type=("build", "run"))
+
+    variant("io", default=True, description="Enable IO features")
+    variant("processing", default=True, description="Enable advanced processing features")
```
It makes it easier to review when the structure conforms to guidelines (see https://spack.readthedocs.io/en/latest/packaging_guide_creation.html#structure-of-a-package).

In this case, move the `variant`s to above the `depends_on` section.

tldahlgren commented on repos/spack_repo/builtin/packages/py_datatrove/package.py:
```diff
@@ -0,0 +1,61 @@
+        depends_on("py-zstandard", type=("build", "run"))
+
+    with when("+processing"):
+        depends_on("py-fasttext-numpy2", type=("build", "run"))
```
Does this replace the following (`0.4.0` and `0.5.0`, https://github.com/huggingface/datatrove/blob/v0.4.0/pyproject.toml)?

```
processing = [
    "fasttext-numpy2-wheel",
...
```

thomas-bouvier replied:
I added the appropriate package. Both seem equivalent.

tldahlgren commented on repos/spack_repo/builtin/packages/py_dateparser/package.py:
```diff
@@ -15,15 +15,19 @@ class PyDateparser(PythonPackage):
 
     variant("calendars", default=True, description="Add calendar libraries")
 
     depends_on("python@2.7:2.8,3.5:", type=("build", "run"))
```
Looks like `1.2.1` requires minimum `python@3.8:` for fuzzing per (https://github.com/scrapinghub/dateparser/blob/02bd2e5dd4477b4f6db98c5e98149458eb3cc821/setup.py#L48).

tldahlgren commented on repos/spack_repo/builtin/packages/py_dateparser/package.py:
```diff
@@ -15,15 +15,19 @@ class PyDateparser(PythonPackage):
 
     depends_on("python@2.7:2.8,3.5:", type=("build", "run"))
     depends_on("py-setuptools", type="build")
     depends_on("py-python-dateutil@2.7.5:", type=("build", "run"))
```
The minimum for `py-python-dateutil` is lower at https://github.com/scrapinghub/dateparser/blob/02bd2e5dd4477b4f6db98c5e98149458eb3cc821/setup.py#L32.

tldahlgren commented on repos/spack_repo/builtin/packages/py_dateparser/package.py:
```diff
@@ -15,15 +15,19 @@ class PyDateparser(PythonPackage):
     depends_on("py-python-dateutil@2.7.5:", type=("build", "run"))
+    depends_on("py-pytz@2024.2:", when="@1.2.1:", type=("build", "run"))
     depends_on("py-pytz", type=("build", "run"))
+    depends_on("py-regex@2024.9.11:", when="@1.2.1:", type=("build", "run"))
```
This is tighter than stated at https://github.com/scrapinghub/dateparser/blob/02bd2e5dd4477b4f6db98c5e98149458eb3cc821/setup.py#L34.

tldahlgren commented on repos/spack_repo/builtin/packages/py_fastapi/package.py:
```diff
@@ -16,6 +16,7 @@ class PyFastapi(PythonPackage):
 
     license("MIT")
 
+    version("0.115.12", sha256="1e2c2a2646905f9e83d32f04a3f86aff4a286669c6c950ca95b5fd68c2602681")
```
I didn't dig into this but am curious as to why the `standard` optional dependencies (see https://github.com/fastapi/fastapi/blob/628c34e0cae200564d191c95d7edea78c88c4b5e/pyproject.toml#L60) are not included?

tldahlgren replied:
Also, the lower bounds for a couple of `+all` are higher for `0.115.12`: `py-jinja2@3.1.5:` (https://github.com/fastapi/fastapi/blob/628c34e0cae200564d191c95d7edea78c88c4b5e/pyproject.toml#L84) and `py-python-multipart@0.0.18:` (https://github.com/fastapi/fastapi/blob/628c34e0cae200564d191c95d7edea78c88c4b5e/pyproject.toml#L86).

thomas-bouvier replied:
I don't know why `standard` optional dependencies are not included

thomas-bouvier replied:
I fixed the version bounds.

tldahlgren commented on repos/spack_repo/builtin/packages/py_fasttext_numpy2/package.py:
```diff
@@ -0,0 +1,24 @@
+    version("0.10.4", sha256="156e84cf2c7db95b24897884284be52c1038fe2b1d0bd9f21bcaf363d2542825")
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
```
There appears to be a lower bound in `setup.py` (same place as the lower bound for `py-pybind11`) -- https://github.com/simon-ging/fasttext-numpy2/blob/baf99646904e34cd2277e63dc0f1be5621c49b73/setup.py#L201 -- of `py-setuptools@0.7:`.

tldahlgren commented on repos/spack_repo/builtin/packages/py_faust_cchardet/package.py:
```diff
@@ -0,0 +1,23 @@
+    version("2.1.19", sha256="f89386297cde0c8e0f5e21464bc2d6d0e4a4fc1b1d77cdb238ca24d740d872e0")
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-wheel", type="build")
```
FWIW. This dependency is inherited from `PythonPackage`.

tldahlgren replied:
There's also a dependency on `python` (which is not inherited from `PythonPackage`), though you'll get the one provided by one of the dependencies (e.g., `py-setuptools`).

thomas-bouvier replied:
Thanks for the clarification.

tldahlgren commented on repos/spack_repo/builtin/packages/py_htmldate/package.py:
```diff
@@ -0,0 +1,29 @@
+    version("1.9.3", sha256="ac0caf4628c3ded4042011e2d60dc68dfb314c77b106587dd307a80d77e708e9")
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
```
There is a lower bound at https://github.com/adbar/htmldate/blob/v1.9.3/pyproject.toml#L3 of `py-setuptools@61:`.

tldahlgren reviewed (changes requested):
I've gone through about half the packages to confirm dependencies. So far I've noticed several discrepancies (see comments) between the package here and the source code (`pyproject.toml` or `setup.*`).

tldahlgren commented on repos/spack_repo/builtin/packages/py_httptools/package.py:
```diff
@@ -21,6 +21,9 @@ class PyHttptools(PythonPackage):
-    depends_on("py-cython@0.29.24:0.29", type="build", when="@:0.6.1")
+    depends_on("py-cython@0.29.24:", when="@0.6.2:", type="build")
+    depends_on("py-cython@0.29.24:0.29", when="@0.5.0", type="build")
+    depends_on("py-cython@0.29.14", when="@0.1.1", type="build")
```
This should be a lower bound (i.e., `py-cython@0.29.14:`).

thomas-bouvier replied:
This answers the question I asked in my last comment, thank you.

tldahlgren commented:
Apologies if I missed it but please confirm you are able to successfully build *every* new version added to the packages on some platform (ideally with `spack debug report` information for the system(s) used.

thomas-bouvier committed:
py-datatrove: address reviews

thomas-bouvier committed:
py-python-multipart: add v0.0.20

thomas-bouvier committed:
py-python-multipart: drop v0.0.5

This version depends on unsupported Python versions.

thomas-bouvier committed:
feat: add package py-fasttext-numpy-wheel

thomas-bouvier committed:
py-fasttext-numpy2-wheel: fix package url

thomas-bouvier commented:
I addressed all reviews, thank you very much for your time @tldahlgren. All dependency versions compile on my machine:

```
% spack debug report
* **Spack:** 1.0.0.dev0 (https://github.com/spack/spack/commit/529e4a7f1a3b09fccfa525c34a87e79805c7fea7)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/85f85c7988e3863826125580e32b084679b9b26e
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-zen3
```

tldahlgren commented on repos/spack_repo/builtin/packages/py_fasttext_numpy2_wheel/package.py:
```diff
@@ -0,0 +1,24 @@
+    """Library for efficient text classification and representation
+    learning, fixed for numpy 2 compatibility."""
+
+    homepage = "https://github.com/hynky1999/Fasttext-np-2.0.0"
```
:thinking:  

This seems like an "odd" package. According to `pypi`, the homepage is https://github.com/facebookresearch/fastText (which links to https://fasttext.cc/). That repository IS marked read-only.

I can (essentially) confirm the dependencies based on that repository *and* the contents of the download.

tldahlgren replied:
I don't see any indications at the URL on line 14 that the repository is necessarily going to replace the one listed in `pypi`. Do you have more information on this matter?

thomas-bouvier replied:
I agree that this is an odd situation @tldahlgren.

1) The original repository is https://github.com/facebookresearch/fastText (not present is Spack). This repository is no longer maintained and is not compatible with Numpy 2+.
2) fastText is a dependency of datatrove, which also depends on Numpy 2+.
3) In this PR, I initially added `py-fasttext-numpy2` https://github.com/simon-ging/fasttext-numpy2, which is a fork of the original fastText with support for Numpy 2+.
4) In this commit I also added `py-fasttest-numpy2` https://github.com/hynky1999/Fasttext-np-2.0.0, which another fork of the original fastText with support for Numpy 2+. I chose this repository because it was uploaded by the same author as `py-datatrove` https://pypi.org/user/hynky/.

Going forward, I can either:

- Keep `py-fasttext-numpy2`
- Keep `py-fasttext-numpy2-wheel`
- Add the original `py-fasttext` and prepare a patch that would be applied with `^numpy@2:`. In this scenario, I could remove `py-fasttext-numpy2` and `py-fasttext-numpy2-wheel`.

wdconinc replied:
`hynky` also the same user as listed in https://pypi.org/project/fasttext-numpy2-wheel/, and it is what `datatrove` would pull in per https://github.com/huggingface/datatrove/blob/main/pyproject.toml#L52, so they are ok with this situation :-)

tldahlgren replied:
@thomas-bouvier @wdconinc Thanks for the answers.

tldahlgren reviewed (commented):
[Re-]confirmed the homepages, checksums, and branch. I have some reservations/a question about the new package's homepage.

tldahlgren commented:
> Apologies if I missed it but please confirm you are able to successfully build _every_ new version added to the packages on some platform (ideally with `spack debug report` information for the system(s) used.

~Please also provide this confirmation/information.~ (🤦 Scrolled through comments and see this was already provided.)

thomas-bouvier commented:
Many thanks @tldahlgren and other reviewers
